cpp/src/iterator/iterator.cuh

```diff
@@ -0,0 +1,313 @@
+  }
+};
+
+/** -------------------------------------------------------------------------*
+ * @overload pair_accessor<ElementType, ResultType, false>
+ * @brief specialization for columns that don't contain null values
+ * -------------------------------------------------------------------------**/
+template <typename ElementType, typename ResultType>
+struct pair_accessor<ElementType, ResultType, false> : public value_accessor<ElementType, ResultType, false>
+{
+/** -------------------------------------------------------------------------*
+ * @brief constructor
+ * @param[in] _col column device view of cudf column
+ * @param[in] i not used
+ * -------------------------------------------------------------------------**/
+  pair_accessor(column_device_view const _col, ResultType i)
+    : value_accessor<ElementType, ResultType, false>(_col, i) {};
+
+  CUDA_DEVICE_CALLABLE
+  thrust::pair<ResultType, bool> operator()(cudf::size_type i) const {
+    return thrust::make_pair(static_cast<ResultType>(this->col.template element<ElementType>(i)), true);
+  }
+};
+} // namespace experimental
+
+// ---------------------------------------------------------------------------
+// helper functions to make iterator
+
+/** -------------------------------------------------------------------------*
+ * @brief Constructs an iterator over the elements of a column.
+ *
+ * If the column contains no null values (indicated by `has_nulls == false`) 
+ * then dereferencing an iterator `it` returned by this function as `*(it + n)` 
+ * will return `ResultType{ static_cast<ElementType*>(data)[n] }`.
+ * 
+ * If the column contains null values (indicated by `has_nulls == true`) 
+ * then the result of de-referencing an iterator `it` returned by this function
+ * as `*(it+n)` will depend if element is valid or null. 
+ * If the element is valid, 
+ * it will return `ResultType{ static_cast<ElementType*>(data)[n] }`.
+ * If the element is null, it will return `ResultType{identity}`.
+ *
+ * @tparam has_nulls Indicates if the column contains null values 
+ *                   (`null_count > 0`)
+ * @tparam ElementType The cudf data type of input array
+ * @tparam ResultType  cudf data type of output and `identity` value
+ *                     which is used when null bitmaps flag is false.
+ * @tparam Iterator_Index
+ *                     The base iterator which gives the index of array.
+ *                     The default is `thrust::counting_iterator`
+ *
+ * @param[in] column   The column device view of cudf column 
+ * @param[in] identity The identity value used when the mask value is false
+ * @param[in] it       The index iterator, `thrust::counting_iterator` by default
+ * -------------------------------------------------------------------------**/
+template <bool has_nulls, typename ElementType, typename ResultType = ElementType,
+    typename Iterator_Index=thrust::counting_iterator<cudf::size_type> >
+auto make_iterator(const column_device_view& column,
```

Comment:
I always felt uneasy about this iterator, and now that I've had time to reflect, I realize why: it does too many things.

It does two things:
1. If nullable, returns the actual value vs. a replacement value (identity)
2. It _also_ casts the returned value to `ResultType`

These things should be decoupled into separate pieces that can be layered together.

To accomplish 1, something like this:
```
/**
 * @brief Constructs an iterator over a column's values that replaces null
 * elements with a specified value.
 *
 * Dereferencing the returned iterator for element `i` will return `column[i]`
 * if it is valid, or `null_replacement` if it is null.
 *
 * @tparam Element The type of elements in the column
 * @param column The column to iterate
 * @param null_replacement The value to return for null elements
 * @return auto Iterator that returns valid column elements, or a null
 * replacement value for null elements.
 */
template <typename Element>
auto make_null_replacement_iterator(column_device_view column,
                                    Element null_replacement = Element{});
```

For 2., I don't think a special libcudf iterator is required, this can be done via a `thrust::transform_iterator` like this:
```
  auto iter = thrust::make_transform_iterator(
      make_null_replacement_iterator(column, identity),
      [] __device__(auto e)(return ResultType{e};));
```

If you find it more convenient, you can wrap this up in another convenience function.

Reply:
There's also no need for the `has_nulls` template parameter. This iterator should only be used in a code path where `column.has_nulls() == true`. Otherwise, `column.begin<Element>()` can be used instead.

Reply:
I felt the same. because of type_dispatcher on 3 template params, there are approximately 150 kernels created for each operator. These variants are created for casting which is not compute/memory intensive, so it felt too much compile time is spent on this.

If `has_nulls` template parameter is removed, it will be ~75 kernels.  
Also, To reduce no. of kernels further, I suggested to move only of typecast into kernel. @harrism expected that kernel would be slower if type dispatch inside kernel. After your suggested simplification, I will try this as well.

Reply:
> There's also no need for the `has_nulls` template parameter. This iterator should only be used in a code path where `column.has_nulls() == true`. Otherwise, `column.begin<Element>()` can be used instead.

Why code path conditional block is preferred over having `has_nulls` as template param?

Reply:
> Why code path conditional block is preferred over having `has_nulls` as template param?

Because there's not much point in having a null replacement iterator if the column doesn't contain any null elements. 

So I'm imagining it would look like this:
```
if(col.has_nulls()){
   auto begin = make_null_replacement_iterator<int32_t>(col,...);
   // do stuff
} else {
   auto begin = col.begin<in32_t>();
   // do stuff 
}
```

Reply:
This point is addressed as separate PR https://github.com/rapidsai/cudf/pull/32981